internal/fans/nvidia.go

```diff
@@ -0,0 +1,228 @@
+}
+
+// helper function to turn an nvml error/return code into a go error
+// (also handles success by returning nil)
+func nvError(errCode nvml.Return) error {
+	if errCode == nvml.SUCCESS {
+		return nil
+	}
+	return errors.New(nvml.ErrorString(errCode))
+}
+
+func (fan *NvidiaFan) getNvFanIndex() int {
+	// seems like fan indices in fan2go are 1-based,
+	// here we start at 0.
+	return fan.Config.Nvidia.Index - 1
+}
+
+func (fan *NvidiaFan) GetId() string {
+	return fan.Config.ID
+}
+
+func (fan *NvidiaFan) GetMinPwm() int {
+	// if the fan is never supposed to stop,
+	// use the lowest pwm value where the fan is still spinning
+	if fan.ShouldNeverStop() {
+		if fan.MinPwm != nil {
+			return *fan.MinPwm
+		}
+	}
+
+	return MinPwmValue
+}
+
+func (fan *NvidiaFan) SetMinPwm(pwm int, force bool) {
+	if fan.Config.MinPwm == nil || force {
+		fan.MinPwm = &pwm
+	}
+}
+
+func (fan *NvidiaFan) GetStartPwm() int {
+	if fan.StartPwm != nil {
+		return *fan.StartPwm
+	} else {
+		return 100
+	}
+}
+
+func (fan *NvidiaFan) SetStartPwm(pwm int, force bool) {
+	if fan.Config.StartPwm == nil || force {
+		fan.StartPwm = &pwm
+	}
+}
+
+func (fan *NvidiaFan) GetMaxPwm() int {
+	if fan.MaxPwm != nil {
+		return *fan.MaxPwm
+	} else {
+		return 100
+	}
+}
+
+func (fan *NvidiaFan) SetMaxPwm(pwm int, force bool) {
+	if fan.Config.MaxPwm == nil || force {
+		fan.MaxPwm = &pwm
+	}
+}
+
+func (fan *NvidiaFan) GetRpm() (int, error) {
+	return fan.GetPwm()
+	// TODO: for driver 565 and newer we can get the actual RPM with custom cgo
+	// code, see my_DeviceGetFanSpeedRPM() in Daniel's nvml test code.
+	// But otherwise, I guess returning the current PWM should also work as it
+	// at least allows creating a fan curve
+}
+
+func (fan *NvidiaFan) GetRpmAvg() float64 {
+	return fan.RpmMovingAvg
+}
+
+func (fan *NvidiaFan) SetRpmAvg(rpm float64) {
+	fan.RpmMovingAvg = rpm
+}
+
+func (fan *NvidiaFan) GetPwm() (int, error) {
+	var device nvml.Device = nil // TODO!
+	fanIdx := fan.getNvFanIndex()
+	speed, ret := device.GetFanSpeed_v2(fanIdx)
+	if ret != nvml.SUCCESS {
+		speed = MinPwmValue // this is what HwMonFan does
+	}
+	// TODO: convert speed from percent (0..100) to PWM (0..255)?
+	return int(speed), nvError(ret)
+}
+
+func (fan *NvidiaFan) SetPwm(pwm int) (err error) {
+	ui.Debug("Setting Fan PWM of '%s' to %d ...", fan.GetId(), pwm)
+	var device nvml.Device = nil // TODO!
+	fanIdx := fan.getNvFanIndex()
+	// TODO: translate pwm (0..255) to percent (0..100)?
+	// or just clamp to 100 and let fan2go assume that it doesn't get faster after PWM > 100?
+	pwm = min(pwm, 100)
+	ret := device.SetFanSpeed_v2(fanIdx, pwm)
+	return nvError(ret)
+}
+
+func (fan *NvidiaFan) GetFanRpmCurveData() *map[int]float64 {
+	return fan.FanCurveData
+}
+
+// AttachFanCurveData attaches fan curve data from persistence to a fan
+// Note: When the given data is incomplete, all values up until the highest
+// value in the given dataset will be interpolated linearly
```

Comment:
You probably copied this from somewhere else, but there is no interpolation logic in here anywere, so I would remove this part.

Reply:
ok